pallets/asset-handler/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+	type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	pub trait Config: frame_system::Config + chainbridge::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Balances Pallet
+		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+        /// Asset Manager
+		type AssetManager: Create<<Self as frame_system::Config>::AccountId>
+		+ Mutate<<Self as frame_system::Config>::AccountId, Balance = u128, AssetId = u128>
+		+ Inspect<<Self as frame_system::Config>::AccountId>;
+
+		/// Asset Create/ Update Origin
+		type AssetCreateUpdateOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		/// Treasury PalletId
+		#[pallet::constant]
+		type TreasuryPalletId: Get<PalletId>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	/// List of relayers who can relay data from Ethereum
+	#[pallet::storage]
+	#[pallet::getter(fn get_bridge_fee)]
+	pub(super) type BridgeFee<T: Config> =
+		StorageMap<_, Blake2_128Concat, BridgeChainId, (BalanceOf<T>, u32), ValueQuery>;
+
+
+	// Pallets use events to inform users when important changes are made.
+	// https://substrate.dev/docs/en/knowledgebase/runtime/events
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Asset Registered
+		AssetRegistered(ResourceId),
+		/// Asset Deposited (recipient, ResourceId, Amount)
+		AssetDeposited(T::AccountId, ResourceId, BalanceOf<T>),
+		/// Asset Withdrawn (recipient, ResourceId, Amount)
+		AssetWithdrawn(H160, ResourceId, BalanceOf<T>),
+		FeeUpdated(BridgeChainId, BalanceOf<T>, )
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Migration is not operational yet
+		NotOperational,
+		/// MinterMustBeRelayer
+		MinterMustBeRelayer,
+		/// ChainIsNotWhitelisted
+		ChainIsNotWhitelisted,
+		/// NotEnoughBalance
+		NotEnoughBalance
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+		/// Creates new Asset where AssetId is derived from chain_id and contract Address
+		///
+		/// # Parameters
+		///
+		/// * `chain_id`: Asset's native chain
+		/// * `contract_add`: Asset's actual address at native chain
+		#[pallet::weight(195_000_000)]
+		pub fn create_asset(
+			origin: OriginFor<T>,
+			chain_id: BridgeChainId,
+			contract_add: H160
+		) -> DispatchResult {
+			T::AssetCreateUpdateOrigin::ensure_origin(origin)?;
+			let rid = chainbridge::derive_resource_id(chain_id, &contract_add.0);
+			T::AssetManager::create(
```

Comment:
Check if asset_id already exists

Reply:
AssetManager::create will return an error if asset id exists. I wrote a test case for it as well.

Reply:
ok